File: anna-setu-frontend-development/lib/api/wallet.ts
/**
 * AnnaSetu — Digital Wallet & Financial Ledger API Client
 */

import { apiClient } from './client'

export type TransactionType =
  | 'TOP_UP'
  | 'DELIVERY_CHARGE'
  | 'PLATFORM_FEE'
  | 'DRIVER_PAYOUT'
  | 'RESERVATION'
  | 'RESERVATION_RELEASE'
  | 'REFUND'
  | 'BONUS'
  | 'ADJUSTMENT'

export type TransactionStatus = 'PENDING' | 'COMPLETED' | 'FAILED' | 'REVERSED'

export interface Wallet {
  id: string
  user_id: string
  owner_type: 'RECEIVER' | 'DRIVER' | 'DONOR' | 'ADMIN'
  balance: number
  reserved_balance: number
  available_balance: number
  currency: 'INR'
  created_at?: string
  updated_at?: string
}

export interface Transaction {
  id: string
  wallet_id: string
  user_id: string
  type: TransactionType
  amount: number
  currency: 'INR'
  status: TransactionStatus
  reference_id?: string
  reference_type?: string
  description: string
  created_at?: string
}

export interface WalletTopUpPayload {
  amount: number
  idempotency_key?: string
}

export interface WalletTopUpResponse {
  wallet_id: string
  amount: number
  new_balance: number
  transaction_id: string
  status: string
  payment_reference?: string
  message: string
}

export interface WalletReservationResponse {
  id: string
  wallet_id: string
  delivery_id: string
  reserved_amount: number
  status: string
  created_at?: string
}

export interface PaymentRefundPayload {
  reason: string
  amount?: number
}

export interface PaymentRefundResponse {
  refund_id: string
  delivery_id: string
  refund_amount: number
  wallet_id: string
  status: string
  message: string
}

export const walletApi = {
  /**
   * Fetch current user's digital wallet and available balance
   */
  async getWallet(): Promise<Wallet> {
    return apiClient<Wallet>('/api/v1/wallet')
  },

  /**
   * Retrieve auditable transaction ledger entries
   */
  async getTransactions(): Promise<Transaction[]> {
    return apiClient<Transaction[]>('/api/v1/wallet/transactions')
  },

  /**
   * Credit wallet balance via demo mode or payment gateway
   */
  async topUp(payload: WalletTopUpPayload): Promise<WalletTopUpResponse> {
    return apiClient<WalletTopUpResponse>('/api/v1/wallet/top-up', {
      method: 'POST',
      body: JSON.stringify(payload),
    })
  },

  /**
   * Reserve delivery total (charge + 12% platform fee) before mission commitment
   */
  async reserveDelivery(deliveryId: string): Promise<WalletReservationResponse> {
    return apiClient<WalletReservationResponse>(`/api/v1/deliveries/${deliveryId}/reserve`, {
      method: 'POST',
    })
  },

  /**
   * Final settlement of completed delivery mission
   */
  async settleDelivery(deliveryId: string): Promise<any> {
    return apiClient<any>(`/api/v1/deliveries/${deliveryId}/settle`, {
      method: 'POST',
    })
  },

  /**
   * Administrative refund for settled delivery
   */
  async refundDelivery(
    deliveryId: string,
    payload: PaymentRefundPayload
  ): Promise<PaymentRefundResponse> {
    return apiClient<PaymentRefundResponse>(`/api/v1/deliveries/${deliveryId}/refund`, {
      method: 'POST',
      body: JSON.stringify(payload),
    })
  },
}

File: backend/app/schemas/wallet.py
"""Wallet and Financial Transaction Schemas."""

from decimal import Decimal
from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field


class TransactionType(str, Enum):
    TOP_UP = "TOP_UP"
    DELIVERY_CHARGE = "DELIVERY_CHARGE"
    PLATFORM_FEE = "PLATFORM_FEE"
    DRIVER_PAYOUT = "DRIVER_PAYOUT"
    RESERVATION = "RESERVATION"
    RESERVATION_RELEASE = "RESERVATION_RELEASE"
    REFUND = "REFUND"
    BONUS = "BONUS"
    ADJUSTMENT = "ADJUSTMENT"


class TransactionStatus(str, Enum):
    PENDING = "PENDING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    REVERSED = "REVERSED"


class WalletResponse(BaseModel):
    id: str
    user_id: str
    owner_type: str = "RECEIVER"
    balance: Decimal = Field(..., description="Total committed wallet balance.")
    reserved_balance: Decimal = Field(default=Decimal("0.00"), description="Amount held for ongoing delivery missions.")
    available_balance: Decimal = Field(..., description="Available unencumbered balance (balance - reserved_balance).")
    currency: str = "INR"
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class TransactionResponse(BaseModel):
    id: str
    wallet_id: str
    user_id: str
    type: TransactionType
    amount: Decimal
    currency: str = "INR"
    status: TransactionStatus
    reference_id: Optional[str] = None
    reference_type: Optional[str] = None
    description: str
    created_at: Optional[str] = None


class WalletTopUpRequest(BaseModel):
    amount: Decimal = Field(..., gt=0, le=50000, description="Amount in INR to credit to wallet.")
    idempotency_key: Optional[str] = Field(default=None, description="Client-supplied idempotency key.")


class WalletTopUpResponse(BaseModel):
    wallet_id: str
    amount: Decimal
    new_balance: Decimal
    transaction_id: str
    status: str
    payment_reference: Optional[str] = None
    message: str = "Wallet top-up processed successfully."


class WalletReservationResponse(BaseModel):
    id: str
    wallet_id: str
    delivery_id: str
    reserved_amount: Decimal
    status: str
    created_at: Optional[str] = None
